File: agent-factory/src/orchestrator.js
// ORCHESTRATOR — long-running supervisor. Reads the queue, decides how many/what
// kind of worker agents are needed, mints + spawns them as child processes,
// monitors them, retires idle ones, records batch stats, runs a self-improvement
// pass, and re-schedules itself based on queue depth. Every decision is logged.
import { spawn } from 'node:child_process';
import { fileURLToPath } from 'node:url';
import path from 'node:path';
import queue from './queue.js';
import * as agents from './agents.js';
import router from './router.js';
import { loadConfig, orchLog } from './util.js';
import { ROOT } from './safepath.js';
import { improve } from './selfimprove.js';
import dashboard from './dashboard.js';

const WORKER = path.join(ROOT, 'src', 'worker.js');

// Run async fn over items with a bounded pool (concurrency).
async function runPool(items, size, fn) {
  const results = [];
  let i = 0;
  const workers = Array.from({ length: Math.max(1, size) }, async () => {
    while (i < items.length) {
      const idx = i++;
      results[idx] = await fn(items[idx], idx);
    }
  });
  await Promise.all(workers);
  return results;
}

// Spawn one worker child process for a task; resolve with its parsed JSON report.
function spawnWorker({ task, model, tier, agentName }) {
  return new Promise((resolve) => {
    const child = spawn('node', [
      WORKER, '--task', task.id, '--model', model.id, '--tier', String(tier), '--agent', agentName,
    ], { cwd: ROOT });
    let out = '', err = '';
    child.stdout.on('data', d => out += d);
    child.stderr.on('data', d => err += d);
    child.on('close', (code) => {
      let report;
      try { report = JSON.parse(out.trim().split('\n').pop()); }
      catch { report = { ok: false, error: `unparseable worker output (code ${code}): ${err || out}` }; }
      resolve({ report, code, pid: child.pid });
    });
  });
}

// Ensure the orchestrator has at least one (idle) worker for each needed family.
function provisionWorkers(claimed, cfg) {
  const byFamily = {};
  for (const t of claimed) {
    const fam = agents.familyFor(t.type);
    (byFamily[fam] ||= new Set()).add(t.type);
  }
  const pool = {}; // family -> [agentDef,...]
  for (const fam of Object.keys(byFamily)) {
    let existing = agents.list().filter(a => a.status !== 'retired' && a.handles.some(h => byFamily[fam].has(h)));
    // mint up to concurrency workers per family so spawns map to distinct agents
    const want = Math.min(cfg.concurrency, byFamily[fam].size + 1);
    while (existing.length < want) {
      const def = agents.mint(fam, [...byFamily[fam]]);
      orchLog(`minted worker`, { agent: def.name, family: fam, handles: def.handles });
      existing.push(def);
    }
    pool[fam] = existing;
  }
  return pool;
}

export async function runBatch() {
  const cfg = loadConfig();
  const batchNo = queue.nextBatchNumber();
  const started_at = new Date().toISOString();
  const claimed = queue.claim(cfg.batch_size, batchNo);

  if (claimed.length === 0) {
    orchLog(`batch #${batchNo}: queue empty, nothing to do`);
    return { processed: 0, batchNo };
  }
  orchLog(`batch #${batchNo}: claimed ${claimed.length} task(s)`, {
    types: claimed.map(t => t.type),
  });

  const pool = provisionWorkers(claimed, cfg);
  const rr = {}; // round-robin index per family

  // Route every task first (decision logged), then execute with bounded concurrency.
  const planned = claimed.map(task => {
    const { model, tier, complexity } = router.pickModel(task, cfg);
    const fam = agents.familyFor(task.type);
    const fleet = pool[fam];
    rr[fam] = (rr[fam] ?? -1) + 1;
    const agentDef = fleet[rr[fam] % fleet.length];
    orchLog(`route`, { task: task.id, type: task.type, complexity, tier, model: model.id, agent: agentDef.name });
    return { task, model, tier, agentName: agentDef.name };
  });

  let cost = 0, succeeded = 0, failed = 0, latSum = 0, latN = 0;
  await runPool(planned, cfg.concurrency, async (p) => {
    agents.touch(p.agentName, { status: 'busy' });
    const { report } = await spawnWorker(p);
    if (report.ok) {
      queue.complete(p.task.id, {
        result: { summary: report.summary, artifacts: report.artifacts, metrics: report.metrics },
        model: report.model, tier: report.tier, cost_usd: report.cost_usd, latency_ms: report.latency_ms,
      });
      cost += report.cost_usd || 0; succeeded++; latSum += report.latency_ms || 0; latN++;
      const def = agents.list().find(a => a.name === p.agentName);
      agents.touch(p.agentName, { status: 'idle', tasks_done: (def?.tasks_done || 0) + 1, last_used_at: new Date().toISOString() });
      orchLog(`done`, { task: p.task.id, agent: p.agentName, cost_usd: report.cost_usd, latency_ms: report.latency_ms, summary: report.summary });
    } else {
      const disp = queue.fail(p.task.id, report.error, { requeue: true, maxAttempts: cfg.retry.max_attempts });
      failed++;
      agents.touch(p.agentName, { status: 'idle' });
      orchLog(`FAILED (${disp})`, { task: p.task.id, agent: p.agentName, error: String(report.error).slice(0, 240) });
    }
  });

  const finished_at = new Date().toISOString();
  const stats = {
    n: batchNo, started_at, finished_at,
    processed: planned.length, succeeded, failed,
    cost_usd: Number(cost.toFixed(6)), avg_latency_ms: latN ? latSum / latN : 0,
  };
  queue.recordBatch(batchNo, stats);

  // Retire idle workers past TTL.
  const retired = agents.retireIdle(cfg.idle_worker_ttl_ms);
  if (retired) orchLog(`retired ${retired} idle worker(s)`);

  // Self-improvement pass (tunes config within bounds).
  const depthAfter = queue.depth();
  const imp = improve({ cfg, stats, queueDepth: depthAfter });
  orchLog(`self-improvement`, { findings: imp.findings.length, changes: imp.changes.map(c => `${c.param}:${c.from}->${c.to}`) });

  dashboard.writeHtml();
  orchLog(`batch #${batchNo} complete`, {
    processed: stats.processed, succeeded, failed, batch_cost_usd: stats.cost_usd, queue_depth: depthAfter,
  });
  return { ...stats, batchNo, queueDepth: depthAfter, improvement: imp };
}

// In-process scheduler: cadence adapts to queue depth (no OS cron involved).
function nextInterval(cfg) {
  const d = queue.depth();
  if (d === 0) return cfg.scheduler_base_interval_ms * 3;   // back off when idle
  if (d > cfg.batch_size * 2) return Math.max(800, cfg.scheduler_base_interval_ms / 2); // speed up under load
  return cfg.scheduler_base_interval_ms;
}

export async function runForever({ maxBatches = Infinity, stopWhenDrained = false } = {}) {
  orchLog('orchestrator starting', { mode: process.env.OPENROUTER_API_KEY ? 'live' : 'stub', maxBatches });
  let n = 0, emptyStreak = 0;
  while (n < maxBatches) {
    const r = await runBatch();
    if (r.processed > 0) { n++; emptyStreak = 0; }
    else { emptyStreak++; }
    if (stopWhenDrained && r.processed === 0) {
      orchLog('queue drained, stopping (stopWhenDrained)');
      break;
    }
    if (n >= maxBatches) break;
    const cfg = loadConfig();
    const wait = nextInterval(cfg);
    orchLog(`scheduler sleeping ${wait}ms (queue depth ${queue.depth()})`);
    await new Promise(res => setTimeout(res, wait));
    if (emptyStreak > 5 && !Number.isFinite(maxBatches)) {
      orchLog('idle for 5 cycles; continuing to wait for new tasks');
      emptyStreak = 0;
    }
  }
  orchLog('orchestrator loop ended', { batches_processed: n });
  return n;
}

export default { runBatch, runForever };

File: agent-factory/orchestrator.py
"""orchestrator.py — the self-managing supervisor.

Responsibilities (the "self-management"):
  * Read the queue and decide HOW MANY workers to run (config.concurrency) and
    WHAT KINDS to mint (from the kinds of pending tasks).
  * Mint workers from the template on demand, spawn them as subprocesses, give
    each a claimed task, and monitor them to completion.
  * Retire finished workers; retire idle minted worker *kinds* no longer needed.
  * Run its own recurring jobs via the in-process scheduler (no OS cron).
  * After each batch, run the self-improvement pass and adopt the new config.
  * Log every decision to logs/ORCHESTRATOR.log and publish runtime state for
    the dashboard.

Modes:
  --demo            drain the current queue, prove ≥2 workers + 1 improvement
                    pass, write the dashboard, and exit.
  --serve           run forever: schedule-driven, adapts cadence to queue depth.
  --seed N          seed sample tasks (+N synthetic) before running.
"""
from __future__ import annotations

import argparse
import json
import subprocess
import sys
import time
import uuid

import core
import factory
import scheduler as sched_mod
import seed_tasks
import self_improve
import task_queue
import dashboard


class Orchestrator:
    def __init__(self) -> None:
        core.ensure_dirs()
        task_queue.init_db()
        self.cfg = core.load_config()
        self.batch = 0
        self.active = {}          # worker_id -> {proc, task_id, kind, started}
        self.minted_kinds = set()
        self.sched = sched_mod.Scheduler()
        self.serving = True
        self._spawned_total = 0

    # --- config -------------------------------------------------------------
    def reload_cfg(self) -> None:
        """Adopt config the self-improvement loop may have rewritten."""
        self.cfg = core.load_config()

    # --- decisions ----------------------------------------------------------
    def decide_capacity(self) -> int:
        """How many workers should be running right now?"""
        counts = task_queue.counts()
        pending = counts["pending"]
        want = min(self.cfg["concurrency"], pending)
        free = want - len(self.active)
        core.orch_log("capacity decision", pending=pending,
                      concurrency=self.cfg["concurrency"],
                      active=len(self.active), spawn=max(0, free))
        return max(0, free)

    def ensure_worker_kind(self, kind: str) -> None:
        """Mint a worker module for `kind` if we don't already have one."""
        if kind not in self.minted_kinds:
            path = factory.mint_worker(kind)
            self.minted_kinds.add(kind)
            core.orch_log("minted worker kind", kind=kind, module=path.name)

    # --- spawning / monitoring ---------------------------------------------
    def spawn_one(self) -> bool:
        """Claim a task and spawn a dedicated worker subprocess for it."""
        worker_id = f"W-{self.batch}-{uuid.uuid4().hex[:6]}"
        task = task_queue.claim_next(worker_id, self.batch)
        if task is None:
            return False
        kind = task["kind"]
        self.ensure_worker_kind(kind)
        module = factory.worker_path(kind)
        proc = subprocess.Popen(
            [sys.executable, str(module),
             "--task-id", str(task["id"]), "--worker-id", worker_id],
            stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True,
        )
        self.active[worker_id] = {
            "proc": proc, "task_id": task["id"], "kind": kind,
            "started": time.time(),
        }
        self._spawned_total += 1
        core.orch_log("spawned worker", worker_id=worker_id, task_id=task["id"],
                      kind=kind, pid=proc.pid)
        return True

    def reap(self) -> None:
        """Collect finished workers and retire them."""
        done_ids = []
        for wid, w in self.active.items():
            ret = w["proc"].poll()
            if ret is None:
                continue
            out, err = w["proc"].communicate()
            summary = {}
            if out.strip():
                try:
                    summary = json.loads(out.strip().splitlines()[-1])
                except json.JSONDecodeError:
                    summary = {"raw": out.strip()[:120]}
            core.orch_log(
                "retired worker", worker_id=wid, task_id=w["task_id"],
                exit=ret, status=summary.get("status", "?"),
                model=summary.get("model", "-"),
                cost_usd=summary.get("cost_usd", 0),
            )
            if err.strip():
                core.orch_log("worker stderr", worker_id=wid,
                              tail=err.strip().splitlines()[-1][:160])
            done_ids.append(wid)
        for wid in done_ids:
            self.active.pop(wid, None)

    def retire_idle_kinds(self) -> None:
        """Drop minted worker modules whose kind has no pending/active work.

        This is the 'retire idle agents' behavior at the kind level: the factory
        does not keep specialized agents around once their backlog is gone.
        """
        pending = task_queue.pending_kinds()
        active_kinds = {w["kind"] for w in self.active.values()}
        for kind in sorted(self.minted_kinds):
            if pending.get(kind, 0) == 0 and kind not in active_kinds:
                if factory.retire_minted(kind):
                    core.orch_log("retired idle worker kind", kind=kind)
                self.minted_kinds.discard(kind)

    # --- runtime state for the dashboard -----------------------------------
    def publish_state(self) -> None:
        core.write_state({
            "ts": core.now_iso(),
            "batch": self.batch,
            "concurrency": self.cfg["concurrency"],
            "complexity_threshold": self.cfg["routing"]["complexity_threshold"],
            "poll_interval_seconds": self.cfg["poll_interval_seconds"],
            "active_workers": [
                {"worker_id": wid, "task_id": w["task_id"], "kind": w["kind"],
                 "age_s": round(time.time() - w["started"], 2)}
                for wid, w in self.active.items()
            ],
            "minted_kinds": sorted(self.minted_kinds),
            "spawned_total": self._spawned_total,
            "queue": task_queue.counts(),
            "scheduler": self.sched.status(),
            "serving": self.serving,
        })

    # --- the main work step -------------------------------------------------
    def drain_step(self) -> None:
        """One pass: spawn up to capacity, reap finished, refresh state."""
        self.reap()
        for _ in range(self.decide_capacity()):
            if not self.spawn_one():
                break
        self.reap()
        self.publish_state()

    def run_batch(self) -> dict:
        """Process the whole current queue as one batch, then self-improve."""
        self.batch += 1
        core.orch_log("batch start", batch=self.batch,
                      pending=task_queue.counts()["pending"],
                      concurrency=self.cfg["concurrency"])
        # keep spawning/reaping until nothing is pending and nothing is active
        while True:
            self.drain_step()
            counts = task_queue.counts()
            if counts["pending"] == 0 and not self.active:
                break
            time.sleep(0.05)
        # self-improvement: review this batch and adopt new knobs
        entry = self_improve.review_and_tune(self.batch)
        self.reload_cfg()
        core.orch_log("self-improvement applied", batch=self.batch,
                      reasons=";".join(entry["reasons"]))
        self.retire_idle_kinds()
        self.publish_state()
        dashboard.render_all()
        return entry

    # --- modes --------------------------------------------------------------
    def run_demo(self) -> None:
        core.orch_log("orchestrator boot", mode="demo")
        entry = self.run_batch()
        # If synthetic overflow left tasks, run a second batch to show adaptation.
        if task_queue.counts()["pending"] > 0:
            self.run_batch()
        core.orch_log("orchestrator demo complete",
                      spawned_total=self._spawned_total,
                      done=task_queue.counts()["done"])

    def serve(self, max_seconds: float | None = None) -> None:
        """Long-running supervised loop driven by the in-process scheduler."""
        core.orch_log("orchestrator boot", mode="serve")
        self.sched.every("drain", self.cfg["poll_interval_seconds"], self.drain_step)
        self.sched.every("improve_if_batch_done", 1.0, self._maybe_close_batch)
        self.sched.every("dashboard", 3.0, dashboard.render_all)
        self.sched.every("adapt_cadence", 2.0, self._adapt_cadence)
        self.batch = 1
        start = time.time()
        try:
            while self.serving:
                self.sched.tick()
                self.publish_state()
                if max_seconds and (time.time() - start) >= max_seconds:
                    core.orch_log("serve time budget reached")
                    break
                time.sleep(0.2)
        except KeyboardInterrupt:
            core.orch_log("orchestrator interrupted — shutting down")
        finally:
            self.serving = False
            self.reap()
            self.publish_state()

    def _maybe_close_batch(self) -> None:
        """In serve mode, close a batch (and self-improve) when the queue drains."""
        counts = task_queue.counts()
        if counts["pending"] == 0 and not self.active and counts["done"] > 0:
            already = self.cfg.get("_last_improved_done")
            if already == counts["done"]:
                return
            entry = self_improve.review_and_tune(self.batch)
            self.reload_cfg()
            self.cfg["_last_improved_done"] = counts["done"]
            self.retire_idle_kinds()
            core.orch_log("self-improvement applied", batch=self.batch,
                          reasons=";".join(entry["reasons"]))
            self.batch += 1

    def _adapt_cadence(self) -> None:
        """Self-scheduling: tighten the drain interval when the queue is deep."""
        pending = task_queue.counts()["pending"]
        base = self.cfg["poll_interval_seconds"]
        interval = max(0.2, base / 2) if pending > self.cfg["concurrency"] else base
        self.sched.set_interval("drain", interval)


def main() -> None:
    ap = argparse.ArgumentParser(description="Agent factory orchestrator")
    ap.add_argument("--demo", action="store_true", help="drain queue once and exit")
    ap.add_argument("--serve", action="store_true", help="run forever")
    ap.add_argument("--seconds", type=float, default=None,
                    help="time budget for --serve")
    ap.add_argument("--seed", type=int, default=None,
                    help="seed sample tasks (+N synthetic) before running")
    args = ap.parse_args()

    if args.seed is not None:
        n = seed_tasks.seed(args.seed)
        core.orch_log("seeded tasks", count=n)

    orch = Orchestrator()
    if args.serve:
        orch.serve(max_seconds=args.seconds)
    else:
        orch.run_demo()

    print(json.dumps({
        "batches": orch.batch,
        "spawned_total": orch._spawned_total,
        "queue": task_queue.counts(),
        "global": task_queue.global_stats(),
    }, indent=2))


if __name__ == "__main__":
    main()
